sqldelight-core/src/main/kotlin/com/squareup/sqldelight/core/android/BuildVariant.kt

```diff
@@ -0,0 +1,33 @@
+package com.squareup.sqldelight.core.android
+
+import com.intellij.psi.PsiDirectory
+import java.util.ArrayList
+
+class BuildVariant(root: PsiDirectory): PsiDirectory by root {
+  internal val name = root.name
+
+  /**
+   * Returns an ordered list of the build variants that this build variant will combine into the
+   * apps source. The list is ordered by precedence, with this build variant overriding any of the
+   * source from the returned build variants.
+   */
+  fun parentVariants(): List<BuildVariant> {
+    if (name == "main") return emptyList()
+
+    val result = ArrayList<BuildVariant>()
+    val indexOfBuildType = name.indexOfFirst { it.isUpperCase() }
```

Comment:
im not in love with this but i really dont want to impose a dependency from the core on gradle which I think would be the only way to accurately get variant dependencies.

Open to other ideas

Reply:
I don't think this module should know anything about variants. It should be a transform from input folders to an output folder only. The Gradle/IDE integration should the concept of variants by assembling the necessary variant folders together to invoke the core.

Reply:
alright let me see how that looks

Reply:
Another reason for that is that I don't think you're restricted to a single word for build types or flavors. You could have `src/shrekMovieDebug/sqldelight/` and `src/shrekTheMusicalDebug/sqldelight`, in theory.

Reply:
Sounds good.

If you can, while making changes in this file, consider what it would looks like to create a command-line interface module which also uses core. This would be needed for supporting Bazel.

Reply:
Right. That was my motivation for doing this original implementation, its only dependency is the folder structure.

Reply:
There is another option for the IDE side which is when compiling the files it also compiles a hidden file which describes a map of input -> output. The build tool will generate that mapping and the IDE will wait to generate code until the file exists.

Reply:
although that means adding a new variant would require a fresh build which kinda sucks

Reply:
A new variant requires a Gradle sync in the IDE which is enough to trigger generation of a file.

Reply:
alright - does that solution feel robust enough or should i consider looking into a solution within the studio apis

Reply:
It's a bit depressing. I'm not sure what it means for some future where a Bazel build wants to use the IDE plugin.

Reply:
i like the idea of thinking about a command line interface instead of gradle/IDE and then solving those problems independently.

something like

```
static void main(String[] args) {
  input = args[0]
  dbPath = args[1]
  packageName = args[2]
}
```

`./java SqlDelight "src/internalDebug/sqldelight" "src/internal/sqldelight src/debug/sqldelight src/debug/sqldelight src/main/sqldelight" "com.sample.app"`

Its at least a place to start. One problem is that this "pushes" the info to the core but in the IDE the core needs to be able to pull the information out - if you command click on an element it has to search upwards to find the table element - which might be in another variant.